// @esri/calcite-components/dist/cjs/it-cf5ce1d8.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
'use strict';

const calciteInputTimePicker_calciteTimePicker_entry = require('./calcite-input-time-picker.calcite-time-picker-4bd65765.js');
require('./index-0130b48b.js');
require('./form-3a8ea930.js');
require('./dom-8530141d.js');
require('./guid-752d9527.js');
require('./resources-5f767c3a.js');
require('./interactive-12c3fa24.js');
require('./browser-1161b139.js');
require('./key-f7bcdea2.js');
require('./label-fc3be0da.js');
require('./component-e5d703f3.js');
require('./loadable-8290f8f3.js');
require('./locale-f24a94d6.js');
require('./observers-b8f934f3.js');
require('./focusTrapComponent-ec75e1ea.js');
require('./math-347bb727.js');
require('./t9n-7ce07cfe.js');
require('./openCloseComponent-b0ebdda0.js');
require('./Validation-9f6156fc.js');

// Italian [it]
var locale = {
  name: 'it',
  weekdays: 'domenica_lunedì_martedì_mercoledì_giovedì_venerdì_sabato'.split('_'),
  weekdaysShort: 'dom_lun_mar_mer_gio_ven_sab'.split('_'),
  weekdaysMin: 'do_lu_ma_me_gi_ve_sa'.split('_'),
  months: 'gennaio_febbraio_marzo_aprile_maggio_giugno_luglio_agosto_settembre_ottobre_novembre_dicembre'.split('_'),
  weekStart: 1,
  monthsShort: 'gen_feb_mar_apr_mag_giu_lug_ago_set_ott_nov_dic'.split('_'),
  formats: {
    LT: 'HH:mm',
    LTS: 'HH:mm:ss',
    L: 'DD/MM/YYYY',
    LL: 'D MMMM YYYY',
    LLL: 'D MMMM YYYY HH:mm',
    LLLL: 'dddd D MMMM YYYY HH:mm'
  },
  relativeTime: {
    future: 'tra %s',
    past: '%s fa',
    s: 'qualche secondo',
    m: 'un minuto',
    mm: '%d minuti',
    h: 'un\' ora',
    hh: '%d ore',
    d: 'un giorno',
    dd: '%d giorni',
    M: 'un mese',
    MM: '%d mesi',
    y: 'un anno',
    yy: '%d anni'
  },
  ordinal: function ordinal(n) {
    return n + "\xBA";
  }
};
calciteInputTimePicker_calciteTimePicker_entry.dayjs.locale(locale, null, true);

exports['default'] = locale;
