// @esri/calcite-components/dist/components/calcite-list-item-group.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client/index.js';
import { c as connectInteractive, u as updateHostInteraction, d as disconnectInteractive, I as InteractiveContainer } from './interactive.js';
import { M as MAX_COLUMNS } from './resources4.js';

const CSS = {
    container: "container",
    heading: "heading",
};

const listItemGroupCss = ":host([disabled]){cursor:default;-webkit-user-select:none;user-select:none;opacity:var(--calcite-opacity-disabled)}:host([disabled]) *,:host([disabled]) ::slotted(*){pointer-events:none}:host{display:flex;flex-direction:column;background-color:var(--calcite-color-foreground-1)}:host([filter-hidden]){display:none}:host([disabled]) ::slotted([calcite-hydrated][disabled]),:host([disabled]) [calcite-hydrated][disabled]{opacity:1}.interaction-container{display:contents}.container{margin:0px;display:flex;flex:1 1 0%;padding:0.75rem;font-size:var(--calcite-font-size--1);font-weight:var(--calcite-font-weight-medium);color:var(--calcite-color-text-2)}.heading{padding:0px}:host([hidden]){display:none}[hidden]{display:none}";
const CalciteListItemGroupStyle0 = listItemGroupCss;

const ListItemGroup = /*@__PURE__*/ proxyCustomElement(class ListItemGroup extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.calciteInternalListItemGroupDefaultSlotChange = createEvent(this, "calciteInternalListItemGroupDefaultSlotChange", 6);
        // --------------------------------------------------------------------------
        //
        //  Private Methods
        //
        // --------------------------------------------------------------------------
        this.handleDefaultSlotChange = () => {
            this.calciteInternalListItemGroupDefaultSlotChange.emit();
        };
        this.disabled = false;
        this.filterHidden = false;
        this.heading = undefined;
    }
    // --------------------------------------------------------------------------
    //
    //  Lifecycle
    //
    // --------------------------------------------------------------------------
    connectedCallback() {
        connectInteractive(this);
    }
    componentDidRender() {
        updateHostInteraction(this);
    }
    disconnectedCallback() {
        disconnectInteractive(this);
    }
    // --------------------------------------------------------------------------
    //
    //  Render Methods
    //
    // --------------------------------------------------------------------------
    render() {
        const { disabled, heading } = this;
        return (h(Host, { key: '72b362e6538638b434eaac9f4295059ab7407d2c' }, h(InteractiveContainer, { key: 'b6a3ad744cfdce9fd2e87b4e66e1ae570998231c', disabled: disabled }, h("tr", { key: '3abe960e3c55dc41c0ee098feb658127795a7204', class: CSS.container }, h("td", { key: '02a7a0797bee8da0ba6f45d8fd45b2285c7710ac', class: CSS.heading, colSpan: MAX_COLUMNS }, heading)), h("slot", { key: 'ed7a1afcaadd1b9071900f5ab6b032aca3a1deb7', onSlotchange: this.handleDefaultSlotChange }))));
    }
    get el() { return this; }
    static get style() { return CalciteListItemGroupStyle0; }
}, [1, "calcite-list-item-group", {
        "disabled": [516],
        "filterHidden": [516, "filter-hidden"],
        "heading": [513]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["calcite-list-item-group"];
    components.forEach(tagName => { switch (tagName) {
        case "calcite-list-item-group":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ListItemGroup);
            }
            break;
    } });
}
defineCustomElement$1();

const CalciteListItemGroup = ListItemGroup;
const defineCustomElement = defineCustomElement$1;

export { CalciteListItemGroup, defineCustomElement };
