// @esri/calcite-components/dist/components/calcite-option-group.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
import { proxyCustomElement, HTMLElement, createEvent, h, Fragment } from '@stencil/core/internal/client/index.js';

const optionGroupCss = ":host{display:block}:host([hidden]){display:none}[hidden]{display:none}";
const CalciteOptionGroupStyle0 = optionGroupCss;

const OptionGroup = /*@__PURE__*/ proxyCustomElement(class OptionGroup extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.calciteInternalOptionGroupChange = createEvent(this, "calciteInternalOptionGroupChange", 6);
        this.disabled = false;
        this.label = undefined;
    }
    handlePropChange() {
        this.calciteInternalOptionGroupChange.emit();
    }
    //--------------------------------------------------------------------------
    //
    //  Render Methods
    //
    //--------------------------------------------------------------------------
    render() {
        return (h(Fragment, { key: '8f313d8117f39af40a697464bc439d380b69263e' }, h("div", { key: 'b04f6bd8f1096b3c87417073638d04b0dec6a649' }, this.label), h("slot", { key: '092e7c21870bda3b7b6f9039e2c49baad0e161f9' })));
    }
    static get watchers() { return {
        "disabled": ["handlePropChange"],
        "label": ["handlePropChange"]
    }; }
    static get style() { return CalciteOptionGroupStyle0; }
}, [1, "calcite-option-group", {
        "disabled": [516],
        "label": [1]
    }, undefined, {
        "disabled": ["handlePropChange"],
        "label": ["handlePropChange"]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["calcite-option-group"];
    components.forEach(tagName => { switch (tagName) {
        case "calcite-option-group":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, OptionGroup);
            }
            break;
    } });
}
defineCustomElement$1();

const CalciteOptionGroup = OptionGroup;
const defineCustomElement = defineCustomElement$1;

export { CalciteOptionGroup, defineCustomElement };
