// @esri/calcite-components/dist/collection/components/option-group/option-group.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
import { Fragment, h } from "@stencil/core";
/**
 * @slot - A slot for adding `calcite-option`s.
 */
export class OptionGroup {
    constructor() {
        this.disabled = false;
        this.label = undefined;
    }
    handlePropChange() {
        this.calciteInternalOptionGroupChange.emit();
    }
    //--------------------------------------------------------------------------
    //
    //  Render Methods
    //
    //--------------------------------------------------------------------------
    render() {
        return (h(Fragment, { key: '8f313d8117f39af40a697464bc439d380b69263e' }, h("div", { key: 'b04f6bd8f1096b3c87417073638d04b0dec6a649' }, this.label), h("slot", { key: '092e7c21870bda3b7b6f9039e2c49baad0e161f9' })));
    }
    static get is() { return "calcite-option-group"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["option-group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["option-group.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "When `true`, interaction is prevented and the component is displayed with lower opacity."
                },
                "attribute": "disabled",
                "reflect": true,
                "defaultValue": "false"
            },
            "label": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": true,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Accessible name for the component."
                },
                "attribute": "label",
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "calciteInternalOptionGroupChange",
                "name": "calciteInternalOptionGroupChange",
                "bubbles": true,
                "cancelable": false,
                "composed": true,
                "docs": {
                    "tags": [{
                            "name": "internal",
                            "text": undefined
                        }],
                    "text": ""
                },
                "complexType": {
                    "original": "void",
                    "resolved": "void",
                    "references": {}
                }
            }];
    }
    static get watchers() {
        return [{
                "propName": "disabled",
                "methodName": "handlePropChange"
            }, {
                "propName": "label",
                "methodName": "handlePropChange"
            }];
    }
}
