// @esri/calcite-components/dist/collection/components/tip-group/tip-group.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
import { h } from "@stencil/core";
/**
 * @deprecated Use the `calcite-carousel` and `calcite-carousel-item` components instead.
 * @slot - A slot for adding `calcite-tip`s.
 */
export class TipGroup {
    constructor() {
        this.groupTitle = undefined;
    }
    render() {
        return h("slot", { key: '284f8e5e3890d264b7131cfcb1ceb7f64f17705c' });
    }
    static get is() { return "calcite-tip-group"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["tip-group.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["tip-group.css"]
        };
    }
    static get properties() {
        return {
            "groupTitle": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The component header text for all nested `calcite-tip`s."
                },
                "attribute": "group-title",
                "reflect": false
            }
        };
    }
}
