// @esri/calcite-components/dist/esm/conditionalSlot-fdf8794e.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
import { f as forceUpdate } from './index-2fab6cb8.js';
import { c as createObserver } from './observers-09e0b0a3.js';

let mutationObserver;
const observerOptions = { childList: true };
/**
 * Helper to set up a conditional slot component on connectedCallback.
 *
 * @param component
 * @deprecated Use `onSlotchange` event with `slotChangeHasAssignedElement` DOM utility instead.
 *
 * ```
 * <slot onSlotchange={(event) => this.mySlotHasElement = slotChangeHasAssignedElement(event)} />}
 * ```
 */
function connectConditionalSlotComponent(component) {
    if (!mutationObserver) {
        mutationObserver = createObserver("mutation", processMutations);
    }
    mutationObserver.observe(component.el, observerOptions);
}
/**
 * Helper to tear down a conditional slot component on disconnectedCallback.
 *
 * @param component
 * @deprecated Use `onSlotchange` event with `slotChangeHasAssignedElement` DOM utility instead.
 *
 * ```
 * <slot onSlotchange={(event) => this.mySlotHasElement = slotChangeHasAssignedElement(event)} />}
 * ```
 */
function disconnectConditionalSlotComponent(component) {
    mutationObserver.unobserve(component.el);
}
function processMutations(mutations) {
    mutations.forEach(({ target }) => {
        forceUpdate(target);
    });
}

export { connectConditionalSlotComponent as c, disconnectConditionalSlotComponent as d };
