// @esri/calcite-components/dist/esm/ro-bd79ecfa.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-design-system/blob/main/LICENSE.md for details.
 * v2.9.0
 */
import { d as dayjs } from './calcite-input-time-picker.calcite-time-picker-e9ab6291.js';
import './index-2fab6cb8.js';
import './form-1928b513.js';
import './dom-acbb9e0d.js';
import './guid-be130516.js';
import './resources-235614b4.js';
import './interactive-fe61fef1.js';
import './browser-07f99cbc.js';
import './key-ca0ab42f.js';
import './label-e5374a88.js';
import './component-e0450ae4.js';
import './loadable-d9ec45c6.js';
import './locale-10002228.js';
import './observers-09e0b0a3.js';
import './focusTrapComponent-a6b1239c.js';
import './math-274e3b1f.js';
import './t9n-b8b9eb22.js';
import './openCloseComponent-1d55c434.js';
import './Validation-3de09d3a.js';

// Romanian [ro]
var locale = {
  name: 'ro',
  weekdays: 'Duminică_Luni_Marți_Miercuri_Joi_Vineri_Sâmbătă'.split('_'),
  weekdaysShort: 'Dum_Lun_Mar_Mie_Joi_Vin_Sâm'.split('_'),
  weekdaysMin: 'Du_Lu_Ma_Mi_Jo_Vi_Sâ'.split('_'),
  months: 'Ianuarie_Februarie_Martie_Aprilie_Mai_Iunie_Iulie_August_Septembrie_Octombrie_Noiembrie_Decembrie'.split('_'),
  monthsShort: 'Ian._Febr._Mart._Apr._Mai_Iun._Iul._Aug._Sept._Oct._Nov._Dec.'.split('_'),
  weekStart: 1,
  formats: {
    LT: 'H:mm',
    LTS: 'H:mm:ss',
    L: 'DD.MM.YYYY',
    LL: 'D MMMM YYYY',
    LLL: 'D MMMM YYYY H:mm',
    LLLL: 'dddd, D MMMM YYYY H:mm'
  },
  relativeTime: {
    future: 'peste %s',
    past: 'acum %s',
    s: 'câteva secunde',
    m: 'un minut',
    mm: '%d minute',
    h: 'o oră',
    hh: '%d ore',
    d: 'o zi',
    dd: '%d zile',
    M: 'o lună',
    MM: '%d luni',
    y: 'un an',
    yy: '%d ani'
  },
  ordinal: function ordinal(n) {
    return n;
  }
};
dayjs.locale(locale, null, true);

export { locale as default };
